### ISSUE TITLE
[Q] middleware & login

### ISSUE THREAD

**Issue by hongz1** [urboifox/nextjs-refreshtoken-demo#1]
I love your approaches and have questions.
It would be much more clear to understand and appreciate if you can post simple example of express api for login, refresh, and profile end-points due to following confusions... 

1. Middleware: I noticed that only refresh request sends stringified cookie via header whereas normal api requests send bearer header. Is this intended approach? why not send token via body when it sends refresh request?
2. Login: 
`
// IMPORTANT you should make the login request, or whatever request that makes the server set a cookie in the browser from the client side
// or make it on the server ex: (server actions), but you would need to parse the cookie from the response and set it yourself
`

quite confused here. let's say I call loginAction (server action) from login client component and loginAction sets cookie (next/header)
Is this same as what you mentioned?

**Comment by urboifox** [comment]
Sorry if those parts confused you a bit.

but regarding your questions,

- In the middleware you can't send the refresh token in the body, that ruins the whole point.
we just send it in the cookies and the server parses it from the cookies.

notice, its the `refresh` token.

- in normal api requests, you send the `access` token.
and the best practice to send that is through `Authorization` header.
that's why we send it as an `Authorization` header.
and the backend parses it from there too.

---

if you wanna login using a server action, you absolutely can.
but you will need to parse the cookies from the incoming request, and set it yourself using the cookies() function

I know it's quite confusing, but the back-end server can't set the `refresh` token in your browser's cookies without you being on the client side.
so either you make the login request from the client, or make it from the server and set the cookie yourself.

I hope i answered your questions, because I really suck at explaining stuff.
but I made you an express server so that you can take a look on what happens there

https://github.com/urboifox/express-refreshtoken-demo

**Comment by hongz1** [comment]
Thank you for the explanations. It helped me to understand.
Basically you are trying to set tokens to http only cookies between api and nextjs server so that client side doesn't do anything about cookie access or modifications. Also, nextjs uses middleware only for managing api requests from server side components.

Codes to be fixed in expressJs
Express > login end-point:
`
should set both access and refresh tokens to response cookie and return status code only. 
Otherwise, it conflicts what you mentions on login page in nextjs. nextjs client component should not do anything about cookie management. However it sets refresh token into response cookie and send access and refresh tokens to response body. So login end-point needs to be corrected in order to let other people clearly understand.
`
I think login api end-point should create two separate response cookies with different timeout.

**Comment by urboifox** [comment]
That's interesting approach, and would totally be great if you are the one working on both, front-end and back-end.

But this repo, and the express one, are useful when the back-end is someone else on your team/project.

If I'm making something full-stack it would be great as you said to set both cookies from the server, or even just use Next's route handlers and server actions.

but in a typical working environment, _at least in my team_ that's the back-end server you will often see.

but anyways feel free to edit or create a PR on any repo.

**Comment by hongz1** [comment]
In my case, I don't use refresh token. Instead, I set 14 days expirations of redis session and put user profile and token into there. if requested ip (redis stores initial ip when sending login request) is different or requested token is different, express delete redis session and send 403 for re-login. if those conditions meet but token is expired, send 401 and client sends token refresh request. when it sends, it updates redis session with new token and send back.

I found your codes because managing token in nextjs between server and client sides was unclear to me.

Anyways, thank you and proxy approach in nextjs helped me what to do with my case.
